### ISSUE TITLE
CUDA_ERROR_UNSUPPORTED_PTX_VERSION (222)

### ISSUE THREAD

**Issue by mrinal-shekhar** [openmm/openmm#3585]
Hi 
I have newly installed Openmm. However, upon using python -m openmm.testInstallation. I get the following response.

 OpenMM Version: 7.7
Git Revision: 130124a3f9277b054ec40927360a6ad20c8f5fa6

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Error computing forces with CUDA platform
4 OpenCL - Successfully computed forces

**CUDA platform error: Error loading CUDA module: CUDA_ERROR_UNSUPPORTED_PTX_VERSION (222)**

Median difference in forces between platforms:

Reference vs. CPU: 6.30514e-06
Reference vs. OpenCL: 6.76294e-06
CPU vs. OpenCL: 8.04607e-07
![Screenshot from 2022-04-26 15-40-28](https://user-images.githubusercontent.com/40069534/165379259-ad17e013-c30f-481c-9560-bee1652733e4.png)

All differences are within tolerance.

Essentially, i am having issues with CUDA and get the error: CUDA_ERROR_UNSUPPORTED_PTX_VERSION (222)**

I have NVIDIA-SMI 470.103.01   Driver Version: 470.103.01   CUDA Version: 11.4   installed. 
Regards
Mrinal

**Comment by peastman** [comment]
This probably means you're using a CUDA toolkit that is incompatible with your driver.  470.x is only compatible with CUDA 11.4 and later toolkits.  See [this table](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#deployment-consideration-forward) from the documentation.

Try this command.

```
conda list | grep cudatoolkit
```

You'll probably find conda has somehow picked the wrong toolkit to install.  Unfortunately, it tends to do that a lot.  Or it may pick an appropriate toolkit, but soon after that, your OS automatically upgrades to a newer driver that is incompatible with what conda has installed.  You can force it to install the 11.4 toolkit with

```
conda install -c conda-forge cudatoolkit=11.4
```

**Comment by decarboxy** [comment]
The other possibility here is that your driver is too *old* for your installed version of the cuda toolkit.  If you're in an environment where you cant upgrade your driver (like a compute cluster), nvidia provides forward compatibility packages.  For example, if you need to use cuda toolkit 11.3 but your driver is too old for it, you can install the `cuda-compat-11.3` package and then add it to your `LD_LIBRARY_PATH` like this: `LD_LIBRARY_PATH=/usr/local/cuda/compat:$LD_LIBRARY_PATH`/

The link in the previous comment also talks about it, but i figured i'd address it more directly here for the next person who finds this ticket trying to get openmm working on a gpu cluster like I did

**Comment by kong1-731** [comment]
I am encountering a similar issue with CUDA_ERROR_UNSUPPORTED_PTX_VERSION (222) when I run an OpenMM script,  but when I test the installation it is working.

<img width="663" alt="Screen Shot 2022-07-27 at 11 52 25 AM" src="https://user-images.githubusercontent.com/70727844/181349928-0f101296-4761-4e53-ac35-59fdb7deecb0.png">

 I upgraded the toolkit to 11.4 and updated the driver to 470.129.06.

Therefore, I am not sure how to fix this issue. Thank you for any guidance!
<img width="516" alt="Screen Shot 2022-07-27 at 11 51 16 AM" src="https://user-images.githubusercontent.com/70727844/181349746-7068fc63-3954-4afc-b060-430992ad2e49.png">
<img width="459" alt="Screen Shot 2022-07-27 at 11 51 37 AM" src="https://user-images.githubusercontent.com/70727844/181349796-57c97ff2-9d21-4dfe-83a5-2058d730ed90.png">
<img width="571" alt="Screen Shot 2022-07-27 at 11 51 53 AM" src="https://user-images.githubusercontent.com/70727844/181349835-5db458c0-a6a7-4d51-8809-7d2abded6c86.png">
.

**Comment by kong1-731** [comment]
^ In case this is relevant for anyone else, I was able to fix this error by updating the driver from 470 to 515.

**Comment by zhoubay** [comment]
> The other possibility here is that your driver is too _old_ for your installed version of the cuda toolkit. If you're in an environment where you cant upgrade your driver (like a compute cluster), nvidia provides forward compatibility packages. For example, if you need to use cuda toolkit 11.3 but your driver is too old for it, you can install the `cuda-compat-11.3` package and then add it to your `LD_LIBRARY_PATH` like this: `LD_LIBRARY_PATH=/usr/local/cuda/compat:$LD_LIBRARY_PATH`/
> 
> The link in the previous comment also talks about it, but i figured i'd address it more directly here for the next person who finds this ticket trying to get openmm working on a gpu cluster like I did

I've met the same problem, I'm working on the cluster with CUDA version=11.0.
![image](https://user-images.githubusercontent.com/44105716/188263413-19b03dac-7f39-44fc-b18a-cc0888dd4c33.png)

I wants to follow your instructions, but only to find that I couldn't find the right way to install the `cuda-compat-11.3`. I've found a CUDA pdf maybe related to this, which tells me to install it throught my apt-get. However, after doing 
`apt-get upgrade`  and `apt-get update`, there's still unable to found the package.

this is the link of the pdf. https://docs.nvidia.com/deploy/pdf/CUDA_Compatibility.pdf
and the guidance of installation is in 3.2.1

![image](https://user-images.githubusercontent.com/44105716/188263503-cc8cc40a-3912-4f6e-bebe-5f2c06650959.png)

I've tried cuda-compat-11-3 or cuda-compat-11.3 or cuda-compat-11, but none works.

could you be so kind to tell me how to install the cuda compat or could you give me some blogs or other materials to do what you described?

Thank you so much !

**Comment by peastman** [comment]
I'm not sure what the post you quoted is talking about.  When installing OpenMM on a cluster with an old driver, you specify what toolkit version to use as described in [the manual](http://docs.openmm.org/latest/userguide/application/01_getting_started.html#installing-openmm).  For example, if your driver only supports 11.0 then you specify `cudatoolkit=11.0`.

**Comment by zhoubay** [comment]
> I'm not sure what the post you quoted is talking about. When installing OpenMM on a cluster with an old driver, you specify what toolkit version to use as described in [the manual](http://docs.openmm.org/latest/userguide/application/01_getting_started.html#installing-openmm). For example, if your driver only supports 11.0 then you specify `cudatoolkit=11.0`.

Thank you for your reply! To my understand, it is impossible for devices of `CUDA version=11.0` to use cudatoolkit higher than `11.0`, is it right? But is there any chance or some patches to let me use `cudatoolkit>11.0`, like `11.3` or something?

Thank you for your reply again!

**Comment by peastman** [comment]
It isn't about the device, it's the driver.  Each driver supports a specific range of toolkits.  If you want to use the 11.3 toolkit, you need to upgrade your driver to a version that supports that toolkit.

**Comment by zhoubay** [comment]
> It isn't about the device, it's the driver. Each driver supports a specific range of toolkits. If you want to use the 11.3 toolkit, you need to upgrade your driver to a version that supports that toolkit.

I got it. Thank you very much!

**Comment by peastman** [comment]
Closing this since I believe the question has been answered.  Feel free to reopen if I missed something.

**Comment by BJWiley233** [comment]
Getting same error use this docker image `FROM nvidia/cuda:11.6.2-devel-ubuntu20.04`

`nvidia-smi` shows:
```
$ nvidia-smi
Sun Nov 20 08:01:27 2022
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
```

**Comment by peastman** [comment]
The error means it's using a newer toolkit than your driver supports.  If you installed OpenMM with conda, then it's probably using a toolkit that conda installed rather than the one you got with apt.  Check

```
conda list | grep cudatoolkit
```

If it's too new, you can install an older one with

```
conda install -c conda-forge cudatoolkit=11.6
```

**Comment by BJWiley233** [comment]
Thanks @peastman yes some packages are "forward" compat but I had to do what you said and made sure that only the conda version was cudatoolkit=11.6 and it worked.

**Comment by andychenbruce** [comment]
I think that conda's `cuda-toolkit` should match the `cuda` field in the output of `nvidia-smi`.

For me I had `cuda-toolkit` a 12.4 but `nvidia-smi` was

```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.30.02              Driver Version: 530.30.02    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
```

So I ran `conda install cuda-toolkit=12.1` which downgraded it then the error changed.

**Comment by MoleOrbitalHybridAnalyst** [comment]
Hi all,

I encountered the same error in my environment.

I might have missed something but I do think I have the compatible driver and cudatoolkit (please see below).

```nvidia-smi``` gives
```
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
```

```conda list | grep cuda-toolkit``` gives
```
cuda-toolkit              12.2.2                        0    nvidia/label/cuda-12.2.2
```

```python -m openmm.testInstallation``` gives
```

OpenMM Version: 8.2
Git Revision: 53770948682c40bd460b39830d4e0f0fd3a4b868

There are 3 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Error computing forces with CUDA platform

CUDA platform error: Error loading CUDA module: CUDA_ERROR_UNSUPPORTED_PTX_VERSION (222)

Median difference in forces between platforms:

Reference vs. CPU: 6.30852e-06

All differences are within tolerance.
```

I appreciate any insights and suggestions.

**Comment by peastman** [comment]
At any point did you install a newer CUDA toolkit than that and run OpenMM with it?  If so, it might have cached kernels compiled with the newer one.  If that's the problem, rebooting should fix it.

**Comment by MoleOrbitalHybridAnalyst** [comment]
> At any point did you install a newer CUDA toolkit than that and run OpenMM with it? If so, it might have cached kernels compiled with the newer one. If that's the problem, rebooting should fix it.

Thanks for the prompt response! I am not sure what the problem was exactly.

I managed to run openmm with CUDA later on by first installing cuda-toolkit=12.0 via conda and then openmm. I noticed that cuda-toolkit was not installed as a dependency if I conda installed openmm first (I do have a system-wide cudatoolkit installed somewhere). I am not sure if that was the reason.
